### PR TITLE
Use a unique producer ID

### PR DESCRIPTION
**Pull request by bkirwi**
This uses the combination of sink and worker ID to uniquely identify a Kafka sink producer, instead of the topic name.

### Motivation

The TODO below mentions that using the topic is unsafe for multiple workers in the Kafka sink. As it happens, #11334 is caused by effectively the same issue: we treat the topic as a unique identifier.

Using the sink id resolves the bug, and using the worker id is meant to future-proof a bit for the world where we run multiple workers in the sink that publish concurrently.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - This uses @philip-stoev's test from the linked issue, updated for recent changes to the syntax.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Fixes a user-facing bug.

### CODE REVIEW

File: src/storage/src/sink/kafka.rs [hunk 1]
```diff
@@ -530,13 +537,9 @@ impl KafkaSinkState {
         // if it makes a big difference
         config.set("queue.buffering.max.ms", &format!("{}", 10));
 
-        if connection.exactly_once {
-            // TODO(aljoscha): this only works for now, once there's an actual
```
Comment (aljoscha) [inline src/storage/src/sink/kafka.rs:534]: :raised_hands:

File: test/testdrive/kafka-duplicate-topic.td [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+$ set schema={"type" : "record", "name" : "test", "fields": [ { "name": "f2", "type": "long" } ] }
+
+$ kafka-create-topic topic=topic0
+
+$ kafka-create-topic topic=topic1
+
+$ kafka-ingest format=avro topic=topic0 schema=${schema} publish=true repeat=1
```
Comment (aljoscha) [inline test/testdrive/kafka-duplicate-topic.td:16]: nit: maybe use different keys here, so `key1` and `key2`, say. You can use `sort-messages=true` on the later verify stanza. Or you could interleave producing to Kafka and verifying the output on the sink.

**Review by aljoscha — APPROVED** [review]
Looks very good! I had one nit that should be easy to resolve, if you want, before merging.

File: test/testdrive/kafka-duplicate-topic.td (hunk 2 above)
Comment (bkirwi) [reply test/testdrive/kafka-duplicate-topic.td:16]: sensible! i've made the records distinct, so we're now checking that we get one from each source.

**Pull request merged**